ocrd_validators/ocrd_validators/page_validator.py

```diff
@@ -2,99 +2,244 @@
 
-def handle_inconsistencies(node, strictness, strategy, report):
+def validate_consistency(node, strictness, strategy, check_baseline, check_coords, report, file_id):
     """
-    Check whether the text results on an element is consistent with its child element text results.
+    Check whether the text results on an element is consistent with its child element text results,
+    and whether the coordinates of an element are fully within its parent element coordinates.
     """
     if isinstance(node, PcGtsType):
-        node = node.get_Page()
+        # top-level (start recursion)
+        node_id = node.get_pcGtsId()
+        node = node.get_Page() # has no .id
     elif isinstance(node, GlyphType):
-        return report
+        # terminal level (end recursion)
+        return True
+    else:
+        node_id = node.id
+    tag = node.original_tagname_
+    log.debug("Validating %s %s", tag, node_id)
+    consistent = True
+    if check_coords or check_baseline:
+        if isinstance(node, PageType):
+            parent = node.get_Border()
+        else:
+            parent = node
+        if parent:
+            parent_points = parent.get_Coords().points
+            node_poly = Polygon(polygon_from_points(parent_points))
+            if (not node_poly.is_valid
+                or node_poly.is_empty
+                or node_poly.bounds[0] < 0
+                or node_poly.bounds[1] < 0
+                or node_poly.length < 4):
+                report.add_error(CoordinateValidityError(tag, node_id, file_id, parent_points))
+                log.info("Invalid coords of %s %s", tag, node_id)
+                consistent = False
+        else:
+            node_poly = None
+    for class_, getter, concatenate_with in _HIERARCHY:
+        if not isinstance(node, class_):
+            continue
+        children = getattr(node, getter)()
+        for child in children:
+            consistent = (validate_consistency(child, strictness, strategy,
+                                               check_baseline, check_coords,
+                                               report, file_id)
+                          and consistent)
+            if check_coords and node_poly:
+                child_tag = child.original_tagname_
+                child_points = child.get_Coords().points
+                child_poly = Polygon(polygon_from_points(child_points))
+                if (not child_poly.is_valid
+                    or child_poly.is_empty
+                    or child_poly.bounds[0] < 0
+                    or child_poly.bounds[1] < 0
+                    or child_poly.length < 4):
+                    report.add_error(CoordinateValidityError(child_tag, child.id, file_id, child_points))
+                    log.info("Invalid coords of %s %s", child_tag, child.id)
+                    consistent = False
+                elif not child_poly.within(node_poly):
+                    # TODO: automatic repair?
```

Comment:
Automatic repair if it's just one or two pixels, you mean? Probably better not to do here, I am not sure whether the "fixing-textequiv-consistency" functionality is wise here.

Reply:
> Automatic repair if it's just one or two pixels, you mean?

I thought of really intersecting with the parent polygon – so naughty children will be restricted by their parents. (This is the opposite strategy of ocrd-segment-repair with sanitize=True.) Just as a simple counter-measure that comes with next to no cost and can be used readily.

> Probably better not to do here, I am not sure whether the "fixing-textequiv-consistency" functionality is wise here.

Why not? As long as it's not the default setting...

Reply:
> Why not? As long as it's not the default setting...

Because I would expect that kind of feature in e.g. `ocrd-segment-repair` but not the PAGE validator. Seems non-trivial to use this in a batch processing setting though I haven't tried. Nothing against fixing per se.

Reply:
I see. I thought of ocrd-segment-repair rather as a provider of non-trivial, logical repairs – whereas this (as well as the `strictness=fix` for textequiv) is a more automatic, pragmatic strategy. Besides, a dedicated processor can only add new annotation into the workspace, whereas the validator is responsible for making all existing annotations conform. What do you think?

Reply:
It makes sense to have these formulaic fixing close to the validation of the form, more so as the input data is changed not amended as you say. If it's simple coordinate fixes you meant with the TODO I'm fine having it in here (though we should change the method signatures accordingly, textequiv-strategy etc). If it's more complex, i.e. more than a few lines, it's better to have a dedicated processor

Reply:
>  If it's simple coordinate fixes you meant with the TODO

It is. Where we now check `child_poly.within(parent_poly)` we would then have to do `child_poly.intersection(parent_poly)`, and write that back to the element via `child.get_Coords().points = points_from_polygon(child_poly.coords)`.

> (though we should change the method signatures accordingly, textequiv-strategy etc)

Again I'd like to preserve backwards compatibility, so if `PageValidator.strictness` and `.strategy` are to be renamed, they at least need an alias kwarg.

For the new options, how about:
- CLI: `--page-coordinate-consistency=poly|fix-poly|baseline|fix-baseline|both|fix-both|off`
- API: `check_coords` `fix_coords` `check_baseline` `fix_baseline`

Or should the API more look like the CLI, or conversely?

Reply:
Point taken on backwards compatibility. I will rename the `page[-_]discipline` to `page-cordinate-consistency` but keep the old labels for textequiv checks for now.

Reply:
Please have a look at https://github.com/OCR-D/core/pull/375/commits/5ca8f7324b94cf0ea85d4afb9234d1f4296e7e96

Reply:
> Please have a look at [5ca8f73](https://github.com/OCR-D/core/commit/5ca8f7324b94cf0ea85d4afb9234d1f4296e7e96)

LGTM. I would probably not have elevated this to _deprecation_, but why not. I'll merge this now.

Reply:
So @kba do you want me to add the simple coordinate fixing as outlined above, or is this for a separate PR? (We'd still have to find a good mnemonic for that on the CLI and API.)

Reply:
I'd prefer a separate PR, I want to merge this and try it out on our GT asap